### PR TITLE
Godel Simple GUI

### PR DESCRIPTION
**Pull request by Jmeyer1292**
A simpler user interface that communicates with the blending server, in much the same way as the original GUI. This GUI is meant to be an operator level interface with only a few simple buttons.

### CODE REVIEW

File: godel_simple_gui/package.xml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0"?>
```
Comment (Jmeyer1292) [inline godel_simple_gui/package.xml:1]: Clean up package.xml

File: godel_simple_gui/src/blending_widget.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+void godel_simple_gui::BlendingWidget::onResetButton()
+{
+  active_state_->onReset(*this);
+}
+
+void godel_simple_gui::BlendingWidget::onOptionsButton()
+{
+  ROS_INFO_STREAM("OPTIONS button pressed");
```
Comment (Jmeyer1292) [inline godel_simple_gui/src/blending_widget.cpp]: At least demoted to DEBUG

File: godel_simple_gui/src/blending_widget.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  options_->show();
+}
+
+void godel_simple_gui::BlendingWidget::onOptionsSave()
+{
+  ROS_INFO_STREAM("Param save requested");
```
Comment (Jmeyer1292) [inline godel_simple_gui/src/blending_widget.cpp]: Demote to DEBUG

File: godel_simple_gui/src/blending_widget.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+void godel_simple_gui::BlendingWidget::changeState(GuiState* new_state)
+{
+  ROS_WARN_STREAM("CHANGING STATE!");
```
Comment (Jmeyer1292) [inline godel_simple_gui/src/blending_widget.cpp]: Delete

File: godel_simple_gui/src/blending_widget.cpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+  active_state_ = new_state;
+  connect(new_state, SIGNAL(newStateAvailable(GuiState*)), this, SLOT(changeState(GuiState*)));
+
+  new_state->onStart(*this);
+}
+
+void godel_simple_gui::BlendingWidget::setButtonsEnabled(bool enabled)
+{
+  ROS_WARN("Enabled: %d", enabled);
```
Comment (Jmeyer1292) [inline godel_simple_gui/src/blending_widget.cpp]: Delete

File: godel_simple_gui/src/options/scan_tool_configuration.cpp [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  ui_->setupUi(this);
+
+  connect(ui_->PushButtonAccept, SIGNAL(clicked()), this, SLOT(accept_changes_handler()));
+  connect(ui_->PushButtonCancel, SIGNAL(clicked()), this, SLOT(cancel_changes_handler()));
+  connect(ui_->PushButtonSave, SIGNAL(clicked()), this, SLOT(save_changes_handler()));
+
+  ui_->ComboBoxQualityMetric->addItems(quality_metric_list);
+}
+
+QStringList godel_simple_gui::ScanPlanConfigWidget::quality_metric_list = QStringList() << "rms" << "ra";
```
Comment (Jmeyer1292) [inline godel_simple_gui/src/options/scan_tool_configuration.cpp]: If we're going to switch from string based identification of scan scoring methods, then the associated methods here need to change as well.

File: godel_simple_gui/src/options/scan_tool_configuration.cpp [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  params_.overlap = ui_->LineEditOverlap->text().toDouble();
+  params_.margin = ui_->LineEditMargin->text().toDouble();
+
+  params_.quality_metric = ui_->ComboBoxQualityMetric->currentText().toStdString();
+  params_.window_width = ui_->LineEditWindowWidth->text().toDouble();
+  params_.min_qa_value = ui_->LineEditMinQAValue->text().toDouble();
+  params_.max_qa_value = ui_->LineEditMaxQAValue->text().toDouble();
+}
+
+int godel_simple_gui::ScanPlanConfigWidget::get_quality_combobox_index()
```
Comment (Jmeyer1292) [inline godel_simple_gui/src/options/scan_tool_configuration.cpp:48]: See above comment about change to enums for quality-metrics

File: godel_simple_gui/src/parameter_window_base.cpp [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+void godel_simple_gui::ParameterWindowBase::show()
+{
+  update_display_fields();
+  QWidget::show();
+}
+
+void godel_simple_gui::ParameterWindowBase::accept_changes_handler()
+{
+  ROS_WARN("Accept button pressed");
```
Comment (Jmeyer1292) [inline godel_simple_gui/src/parameter_window_base.cpp]: delete

File: godel_simple_gui/src/parameter_window_base.cpp [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+void godel_simple_gui::ParameterWindowBase::cancel_changes_handler()
+{
+  hide();
+}
+
+void godel_simple_gui::ParameterWindowBase::save_changes_handler()
+{
+  ROS_WARN("Save button pressed");
```
Comment (Jmeyer1292) [inline godel_simple_gui/src/parameter_window_base.cpp]: delete

File: godel_simple_gui/src/states/error_state.cpp [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+#include <QMessageBox>
+
+godel_simple_gui::ErrorState::ErrorState(const std::string& msg, GuiState* next_state)
+  : msg_(msg)
+  , next_state_(next_state)
+{}
+
+void godel_simple_gui::ErrorState::onStart(BlendingWidget& gui)
+{
+  ROS_INFO_STREAM("ErrorState start");
```
Comment (Jmeyer1292) [inline godel_simple_gui/src/states/error_state.cpp]: Delete

File: godel_simple_gui/src/states/error_state.cpp [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  gui.setButtonsEnabled(false);
+
+  // Show error dialog
+  window_ = new QWidget();
+
+  QMessageBox::StandardButton reply;
+  reply = QMessageBox::critical(window_, tr("Warning"), QString::fromStdString(msg_), QMessageBox::Ok);
+//                                tr("QMessageBox::critical()"),
+//                                QString::fromStdString(msg_),
+//                                QMessageBox::Abort | QMessageBox::Retry | QMessageBox::Ignore);
```
Comment (Jmeyer1292) [inline godel_simple_gui/src/states/error_state.cpp]: Delete comments

File: godel_simple_gui/src/states/wait_for_execute_state.cpp [hunk 12]
```diff
@@ -0,0 +1 @@
+
```
Comment (Jmeyer1292) [inline godel_simple_gui/src/states/wait_for_execute_state.cpp:1]: Misnamed file?

File: godel_simple_gui/include/godel_simple_gui/options/surface_detection_configuration.h [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  Q_OBJECT
+public:
+  SurfaceDetectionConfigWidget(godel_msgs::SurfaceDetectionParameters params);
+
+  godel_msgs::SurfaceDetectionParameters& params() { return params_; }
+  const godel_msgs::SurfaceDetectionParameters& params() const { return params_; }
+
+  virtual void update_display_fields();
+
+  virtual void update_internal_fields();
```
Comment (jrgnicho) [inline godel_simple_gui/include/godel_simple_gui/options/surface_detection_configuration.h:26]: There are several methods in this PR that are named with lowercase and underscores as opposed to the ros guidelines .  As this is the case in many other packages in this repo we should address this in a separate PR as it appears to require a rather large set of changes.

File: godel_simple_gui/include/godel_simple_gui/states/simulating_state.h [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+  virtual void onStart(BlendingWidget& gui);
+  virtual void onExit(BlendingWidget& gui);
+
+  // Handlers for the fixed buttons
+  virtual void onNext(BlendingWidget& gui);
+  virtual void onBack(BlendingWidget& gui);
+  virtual void onReset(BlendingWidget& gui);
+
+  void simulateAll();
+  void simulateOne(const std::string& plan);
```
Comment (jrgnicho) [inline godel_simple_gui/include/godel_simple_gui/states/simulating_state.h:28]: Should these two simulate methods be _protected_?

File: godel_simple_gui/src/blending_widget.cpp [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+  connect(options_, SIGNAL(saveRequested()), this, SLOT(onOptionsSave()));
+
+  // Connect to ROS services
+  loadParameters();
+}
+
+godel_simple_gui::BlendingWidget::~BlendingWidget()
+{
+  delete active_state_;
+  delete options_;
```
Comment (jrgnicho) [inline godel_simple_gui/src/blending_widget.cpp:41]: I believe that Qt widgets are garbage collected so you may not need to explicitly delete these pointers

File: godel_simple_gui/src/blending_widget.cpp [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+void godel_simple_gui::BlendingWidget::onOptionsButton()
+{
+  options_->show();
+}
+
+void godel_simple_gui::BlendingWidget::onOptionsSave()
+{
+}
```
Comment (jrgnicho) [inline godel_simple_gui/src/blending_widget.cpp:71]: Is this empty method implemented in order to adhere to an interface?

File: godel_simple_gui/include/godel_simple_gui/options/surface_detection_configuration.h (hunk 13 above)
Comment (Jmeyer1292) [inline godel_simple_gui/include/godel_simple_gui/options/surface_detection_configuration.h:26]: +1. There are a lot of legacy, if you will, components in Godel that don't follow the ROS conventions. The only sane thing to do is update them independently of changing logic.

File: godel_simple_gui/include/godel_simple_gui/states/simulating_state.h (hunk 14 above)
Comment (Jmeyer1292) [inline godel_simple_gui/include/godel_simple_gui/states/simulating_state.h:28]: +1 I think that makes more sense than public.

File: godel_simple_gui/src/options/scan_tool_configuration.cpp [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+{
+  switch (params_.quality_metric)
+  {
+  case godel_msgs::ScanPlanParameters::METHOD_RMS:
+    return 0;
+  default:
+    return -1;
```
Comment (jrgnicho) [inline godel_simple_gui/src/options/scan_tool_configuration.cpp:55]: Unless more flags are added eventually, using a _switch_ statement here seems overkill.

File: godel_simple_gui/src/options/scan_tool_configuration.cpp [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+int godel_simple_gui::ScanPlanConfigWidget::get_scan_method_enum_value()
+{
+  switch (ui_->ComboBoxQualityMetric->currentIndex())
+  {
+  case 0:
+    return godel_msgs::ScanPlanParameters::METHOD_RMS;
+  default:
+    return -1;
+  }
```
Comment (jrgnicho) [inline godel_simple_gui/src/options/scan_tool_configuration.cpp:67]: Same as the comment above

File: godel_simple_gui/src/states/executing_state.cpp [hunk 19]
```diff
@@ -0,0 +1,92 @@
+#include "godel_simple_gui/states/executing_state.h"
+// prev
+#include "godel_simple_gui/states/wait_to_execute_state.h"
+ //next
+//#include "godel_simple_gui/states/wait_to_execute_state.h"
```
Comment (jrgnicho) [inline godel_simple_gui/src/states/executing_state.cpp]: Remove?

File: godel_simple_gui/src/blending_widget.cpp (hunk 15 above)
Comment (Jmeyer1292) [inline godel_simple_gui/src/blending_widget.cpp:41]: My understanding is that Qt will delete trees of nodes that are connected to one another via the `parent` argument you see in many cases. Note that `ui_` takes `this` as an argument. This is handled through some kind of signal magic. 

So when you don't give it a parent, you should still manually delete it if you don't want to leak memory.

File: godel_simple_gui/src/blending_widget.cpp (hunk 16 above)
Comment (Jmeyer1292) [inline godel_simple_gui/src/blending_widget.cpp:71]: The save parameters in this simple gui are unimplemented. It's more of a "did not finish development" than an intentional omission.

File: godel_simple_gui/src/options/scan_tool_configuration.cpp (hunk 17 above)
Comment (Jmeyer1292) [inline godel_simple_gui/src/options/scan_tool_configuration.cpp:55]: More flags will be added. I've got support for two more methods in the scan analysis package already.

File: godel_simple_gui/src/states/executing_state.cpp (hunk 19 above)
Comment (Jmeyer1292) [inline godel_simple_gui/src/states/executing_state.cpp]: +1

File: godel_simple_gui/src/blending_widget.cpp (hunk 16 above)
Comment (jrgnicho) [inline godel_simple_gui/src/blending_widget.cpp:71]: +1

File: godel_simple_gui/src/options/scan_tool_configuration.cpp (hunk 17 above)
Comment (jrgnicho) [inline godel_simple_gui/src/options/scan_tool_configuration.cpp:55]: +1

File: godel_simple_gui/src/blending_widget.cpp (hunk 15 above)
Comment (jrgnicho) [inline godel_simple_gui/src/blending_widget.cpp:41]: +1

**Comment by jrgnicho** [comment]
+1.  You can merge

**Pull request merged**